### PR TITLE
modify match planner to support tag filter

### PR DESCRIPTION
**Pull request by Aiee**
Add support for MATCH to filter vertex given only vertex tag

relative issues:
NG-316 http://jira.nebula-graph.io/browse/NG-316
NG-317 http://jira.nebula-graph.io/browse/NG-317 (tests skipped)
NG-344 http://jira.nebula-graph.io/browse/NG-344

update:
currently tests containing multi-steps where the node has only tag given are skipped

### CODE REVIEW

**Review by yixinglu — COMMENTED** [review]
Please Add test cases

File: src/validator/MatchValidator.h [hunk 1]
```diff
@@ -41,7 +41,8 @@ class MatchValidator final : public TraversalValidator {
     Status validateStepRange(const MatchStepRange *range) const;
 
     StatusOr<Expression*> makeSubFilter(const std::string &alias,
-                                        const MapExpression *map) const;
+                                        const MapExpression *map,
+                                        const std::string) const;
```
Comment (yixinglu) [inline src/validator/MatchValidator.h:45]: `const std::string &label = ""`

File: src/validator/MatchValidator.cpp [hunk 2]
```diff
@@ -384,7 +388,24 @@ Status MatchValidator::validateStepRange(const MatchStepRange *range) const {
 
 StatusOr<Expression*>
 MatchValidator::makeSubFilter(const std::string &alias,
-                              const MapExpression *map) const {
+                              const MapExpression *map,
+                              const std::string label) const {
```
Comment (yixinglu) [inline src/validator/MatchValidator.cpp:392]: `const std::string& label`

File: src/validator/MatchValidator.cpp [hunk 3]
```diff
@@ -384,7 +388,24 @@ Status MatchValidator::validateStepRange(const MatchStepRange *range) const {
...
+
+        auto* args = new ArgumentList();
+        args->addArgument(std::make_unique<LabelExpression>(alias));
+        auto *right = new FunctionCallExpression(
+                    new std::string("tags"),
+                    args);
+        root = new RelationalExpression(Expression::Kind::kRelIn, left, right);
+
+        saveObject(root);
+        return root;
```
Comment (yixinglu) [inline src/validator/MatchValidator.cpp:406]: `return saveObject(root)`

File: src/validator/MatchValidator.cpp [hunk 4]
```diff
@@ -384,7 +388,24 @@ Status MatchValidator::validateStepRange(const MatchStepRange *range) const {
 
 StatusOr<Expression*>
 MatchValidator::makeSubFilter(const std::string &alias,
-                              const MapExpression *map) const {
+                              const MapExpression *map,
+                              const std::string label) const {
+    // Node has tag without property
+    if (!label.empty() && map == nullptr) {
+        Expression *root = nullptr;
```
Comment (yixinglu) [inline src/validator/MatchValidator.cpp:395]: Need not to declare `root` variable in the beginning of code blocks

File: src/validator/MatchValidator.cpp [hunk 5]
```diff
@@ -175,7 +179,7 @@ Status MatchValidator::buildEdgeInfo(const MatchPath *path,
         }
         Expression *filter = nullptr;
         if (props != nullptr) {
-            auto result = makeSubFilter(*alias, props);
+            auto result = makeSubFilter(*alias, props, "");
```
Comment (yixinglu) [inline src/validator/MatchValidator.cpp:182]: Need not to update it here if you set the default parameter for `makeSubFilter` function

File: src/validator/MatchValidator.cpp [hunk 6]
```diff
@@ -123,7 +123,11 @@ Status MatchValidator::buildNodeInfo(const MatchPath *path,
...
```
Comment (yixinglu) [inline src/validator/MatchValidator.cpp:126]: See following comment

File: src/planner/match/MatchClausePlanner.cpp [hunk 7]
```diff
@@ -145,9 +145,19 @@ Status MatchClausePlanner::appendFetchVertexPlan(const Expression* nodeFilter,
...
+            // filter prop
+            if (expr->kind() == Expression::Kind::kLabelAttribute) {
+                auto la = static_cast<const LabelAttributeExpression*>(expr);
+                res = new AttributeExpression(
+                    new VertexExpression(), la->right()->clone().release());
+                return res;
+            }
+            // filter tag
+            res =  new VertexExpression();
+            return res;
```
Comment (yixinglu) [inline src/planner/match/MatchClausePlanner.cpp:160]:
```cpp
return new VertexExpression();
```

File: src/planner/match/MatchClausePlanner.cpp [hunk 8]
```diff
@@ -145,9 +145,19 @@ Status MatchClausePlanner::appendFetchVertexPlan(const Expression* nodeFilter,
...
-            return new AttributeExpression(new VertexExpression(), la->right()->clone().release());
+            Expression* res = nullptr;
+            DCHECK(expr->kind() == Expression::Kind::kLabelAttribute ||
+                expr->kind() == Expression::Kind::kLabel);
+            // filter prop
+            if (expr->kind() == Expression::Kind::kLabelAttribute) {
+                auto la = static_cast<const LabelAttributeExpression*>(expr);
+                res = new AttributeExpression(
+                    new VertexExpression(), la->right()->clone().release());
+                return res;
```
Comment (yixinglu) [inline src/planner/match/MatchClausePlanner.cpp:156]:
```cpp
return new AttributeExpression(...);
```

File: src/planner/match/Expand.cpp [hunk 9]
```diff
@@ -128,9 +128,19 @@ Status Expand::expandStep(const EdgeInfo& edge,
...
+            // filter prop
+            if (expr->kind() == Expression::Kind::kLabelAttribute) {
+                auto la = static_cast<const LabelAttributeExpression*>(expr);
+                res = new AttributeExpression(
+                    new VertexExpression(), la->right()->clone().release());
+                return res;
+            }
+            // filter tag
+            res =  new VertexExpression();
+            return res;
```
Comment (yixinglu) [inline src/planner/match/Expand.cpp:143]: see following comment

File: tests/tck/features/match/Base.feature [hunk 10]
```diff
@@ -35,3 +35,70 @@ Feature: Basic match
...
+  @skip
+  Scenario: [2] one step given tag without property
+    When executing query:
+      """
+      MATCH (v:team{name:"Spurs"})--(v2)
+      RETURN v2
+      """
+    Then the result should be, in any order:
+      | "v2"                                                               |
+      | {"name":"Boris Diaw","age":36,"player":"Boris Diaw"}               |
```
Comment (yixinglu) [inline tests/tck/features/match/Base.feature:59]: Please rewrite the expected vertex result by the format:

```text
("Boris Diaw" :player{name: "Boris Diaw", age: 36})
```

File: tests/tck/features/match/Base.feature [hunk 11]
```diff
@@ -35,3 +35,70 @@ Feature: Basic match
...
+      | {"name":"Marco Belinelli","age":32,"player":"Marco Belinelli"}     |
+      | {"name":"Dejounte Murray","age":29,"player":"Dejounte Murray"}     |
+
+  @skip
+  Scenario: multi steps given tag without property
+    Given a graph with space named "nba"
+    When executing query:
+      """
+      MATCH (v:player{name: "Tim Duncan"})-->(v2:team)<--(v3)
+      RETURN v2
```
Comment (yixinglu) [inline tests/tck/features/match/Base.feature:84]: Add following cases:

```
      MATCH (v:player{name: "Tim Duncan"})--(v2:team)--(v3)
      RETURN v2
```

File: src/validator/MatchValidator.cpp [hunk 12]
```diff
@@ -384,18 +388,32 @@ Status MatchValidator::validateStepRange(const MatchStepRange *range) const {
...
 MatchValidator::makeSubFilter(const std::string &alias,
-                              const MapExpression *map) const {
+                              const MapExpression *map,
+                              const std::string& label) const {
+    // Node has tag without property
+    if (!label.empty() && map == nullptr) {
+        auto *left = new ConstantExpression(label);
+
+        auto* args = new ArgumentList();
+        args->addArgument(std::make_unique<LabelExpression>(alias));
```
Comment (jievince) [inline src/validator/MatchValidator.cpp:398]: std::make_unique<LabelExpression>(new std::string(alias))

Comment (jievince) [inline src/validator/MatchValidator.cpp:398]: std::make_unique(new std::string(alias))

**Review by CPWstatic — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Pull request merged**